### PR TITLE
Non-record: 11L Depth Recurrence + High-Yield Legal TTT (1.14458 BPB)

### PR DESCRIPTION
**Pull request by Christopher-Lee-McClendon**
## Non-Record Submission: 11L Depth Recurrence + Legal Score-First TTT

**val_bpb = 1.14458** | Pre-TTT: 1.1611 | TTT gain: **−0.0165** | Artifact: 14.79 MB

> Non-record unlimited-compute submission (trained on 4×A100-40GB).

---

### Headline

This submission demonstrates that **competition-legal test-time training can deliver large gains** when properly tuned. The key finding is a TTT recipe — SGD with momentum, multiple epochs per chunk, freezing early blocks — that extracts **2.4× more improvement** (−0.0165 BPB) than single-epoch AdamW over the full network (−0.0068 in our prior PR #456).

Every validation token is **scored before any weight update** that could use it, enforced by `torch.inference_mode()` during scoring.

### Novel & Creative Contributions

1. **High-yield legal TTT via selective freezing + SGD momentum** — Most TTT approaches use AdamW for 1 epoch over all parameters. We use SGD+momentum(0.9) for 3 epochs per 32K chunk while freezing the first 2 blocks. This is simpler, uses less memory, and gets 2.4× better TTT gains.

2. **Depth recurrence** — 11 logical layers from 10 unique BlockCores (one core reused at two depths with independent normalization). Delivers 11-layer capacity at 10-layer parameter cost.

3. **Partial RoPE (16/64 dims)** — Only 16 of 64 head dimensions use rotary embeddings, with NTK-aware scaling. The remaining 48 act as position-agnostic content channels, improving length generalization during TTT.

4. **Value Embeddings on deep layers only** — 128-dim learned embeddings added to value projections on layers 9–10, giving deep layers direct token-identity access in the value stream.

5. **Layer-Norm depth scaling** — `1/√(layer+1)` scaling stabilizes training under depth recurrence.

### Architecture

| Component | Setting |
|---|---|
| Layers | 11 logical (10 shared BlockCores) |
| Dim/Heads | 512 / 8 (4 KV heads) |
| MLP | 3× (1536), ReLU²; SmearGate |
| BigramHash | 2048 |
| RoPE | 16/64 dims, NTK scaling |
| Value Embed | 128d on layers 9–10 |
| XSA | Last 4 layers |
| Quant | Int6 + zstd |

### TTT Protocol

```
for each 32K chunk:
    1. model.eval() + inference_mode → score chunk (NLL recorded)
    2. model.train() → SGD(lr=0.002, mom=0.9), 3 epochs, freeze blocks 0-1
```

### Size Budget

| Component | Bytes |
|---|---|
| Model (int6+zstd) | 14,717,713 |
| Code | 71,706 |
| **Total** | **14,789,419** (< 16,000,000) |

### vs. Prior Submission (PR #456)

| | PR #456 | This PR | Δ |
|---|---|---|---|
| val_bpb | 1.15321 | **1.14458** | **−0.00863** |
| Pre-TTT | 1.1600 | 1.1611 | +0.0011 |
| TTT gain | −0.0068 | **−0.0165** | **2.4×** |

Pre-TTT baselines are nearly identical — the entire improvement comes from better TTT, validating the recipe.

### Credits

This submission builds on work from many contributors to the parameter-golf competition:

- **Muon optimizer** — Baseline (modded-nanogpt); Newton-Schulz orthogonal preconditioning
- **BigramHash embeddings** — PR #65 (aquariouseworkman)
- **SmearGate** — PR #65 (aquariouseworkman)
- **XSA** — PR #187 (Idan3011); GQA-aware variant in PR #265 (unnir)
- **U-Net skip connections** — PR #65 (aquariouseworkman), PR #69 (TevBenji)
- **Mixed int5/int6 quantization** — PR #76 (unixmadtoonslab / Will DePue)
- **SWA** — PR #69 (TevBenji)
- **Late QAT** — PR #315 (jfprincz), PR #374 (unnir)
- **Sliding window evaluation** — PR #50 (mattqlf / Matthew Li)
- **Value Embeddings** — PR #374 (unnir)
- **Partial RoPE** — PR #315 (jfprincz), PR #374 (unnir)
- **LN Scale** — PR #315 (jfprincz), PR #374 (unnir)
- **Depth recurrence** — PR #8 (iranzithierry)
- **Legal TTT framework** — PR #77 (samacqua); full-model variant in our PR #456
- **ReLU², GQA** — Baseline (modded-nanogpt)

Built on the [parameter-golf](https://github.com/openai/parameter-golf) starter code by Beren Millidge & Keller Jordan.

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-record: 11L Depth Recurrence + High-Yield Legal TTT (1.14458 BPB)

**BPB:** 1.14458 | **Compliance:** LOOKS CLEAN — score-first-per-chunk TTT (legal #1416/#1423 pattern)

**What I found in the code** (head SHA `f8ff8039ca8e`, file `records/track_non_record_16mb/2026-03-22_11L_VE128_PartialRoPE_LegalTTT/train_gpt.py`):

The TTT path at line **842** implements the score-first-per-chunk pattern: each chunk is scored under `torch.no_grad()` / `inference_mode()` before the `base_model.train()` + SGD adaptation runs on that same chunk, with an `is_last_chunk` guard so the final chunk gets no adaptation pass. This is the structural shape the legal frontier uses (PRs #1416 erichroepke, #1423 aryanbhosale).

Per **Issue #402 and Issue #677**, TTT is legal when each token is scored before the adapter updates on it, and that's what the code does here — chunk `ci` is scored under weights adapted only on chunks `0..ci-1`. No `prequant_ttt_adapt_adamw(val_tokens, ...)` multi-epoch fine-tune, no scored-region SLOT, no target-in-key n-gram cache.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.21s, dim=512, layers=11, vocab=1024, code=71738 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending standard checks (3-seed validation, 16MB artifact cap, 10-min wallclock on 8×H100 SXM). The compliance picture matches the legal reference frontier and no flags were raised by the classification pass.

*Auto-classification caveat:* this review was drafted by the AST-based classifier against a template derived from manually-reviewed cluster PRs (#1420, #1450, #1487, #1541, #1529, #1533, #1518). If I've misread a subtlety in your eval path — e.g., multi-epoch TTT that I mistook for single-pass, or a target-in-key lookup I missed in a helper function — please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.21s, dim=512, layers=11, vocab=1024, code=71738 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
